feat: Add support for connecting via a hostname instead of IP

ShadauxCat opened the pull request:
Add support for connecting via a hostname instead of IP when using supported versions of unity and unity transport.

## Changelog

- Added: When using UnityTransport >=2.4 and Unity >= 6000.1.0a1, `SetConnectionData` will accept a fully qualified hostname instead of an IP as a connect address on the client side.

## Testing and Documentation

- No tests have been added. (Manual tests were performed. The CI environment doesn't have "localhost" consistently defined to be able to run automated tests.)
- Includes edits to existing public API documentation.

## Backport

Backport: https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/3440

ShadauxCat committed:
Add support for connecting via a hostname instead of IP when using supported versions of unity and unity transport.

ShadauxCat committed:
Changelog

ShadauxCat committed:
Fix failing tests

ShadauxCat committed:
Removed test that is no longer valid.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Editor/Transports/UnityTransportTests.cs:
Is there no longer any way to know if an endpoint is invalid with this update?
Just wondering why we are deleting tests as opposed to updating them.

ShadauxCat replied:
Now that FQDNs are supported, there are very few addresses that would be invalid. We could check for things like whitespace or non-printable characters, but that's checking for things the user's unlikely to provide us. The only way to really know if an endpoint is valid is to try to connect to it, and I'd argue that having the same error (failure to connect) for a typo that misspells .com as .con as we do for a typo that adds a space is more consistent behavior.

But there's no way to make StartClient fail because that check is async, and that's what those tests were testing.

NoelStephensUnity replied:
So my question is more relating to removing something like this:
```
        [Test]
        public void UnityTransport_StartClientFailsWithBadAddress()
        {
            UnityTransport transport = new GameObject().AddComponent<UnityTransport>();
            transport.Initialize();

            transport.SetConnectionData("foobar", 4242);
            Assert.False(transport.StartClient());

            LogAssert.Expect(LogType.Error, "Invalid network endpoint: foobar:4242.");
            LogAssert.Expect(LogType.Error, "Target server network address (foobar) is Invalid!");

            transport.Shutdown();
        }
```
As opposed to just wrapping the part I think you are thinking is no longer valid?
```
        [Test]
        public void UnityTransport_StartClientFailsWithBadAddress()
        {
            UnityTransport transport = new GameObject().AddComponent<UnityTransport>();
            transport.Initialize();

            transport.SetConnectionData("foobar", 4242);
            Assert.False(transport.StartClient());

#if !HOSTNAME_RESOLUTION_AVAILABLE && !UTP_TRANSPORT_2_4_ABOVE
            LogAssert.Expect(LogType.Error, "Invalid network endpoint: foobar:4242.");
            LogAssert.Expect(LogType.Error, "Target server network address (foobar) is Invalid!");
#endif
            transport.Shutdown();
        }
```
Where we would still expect the same behavior of the client failing to start if it passed in just "foobar" as the address... which is not FQDN nor an IP... ?

The same would go for `DetectInvalidEndpoint`.

I am confused about the async part as both `UnityTransport.StartClient` and `UnityTransport.StartServer` are non-async methods that are defined within `NetworkTransport`?

Don't we want to make sure that if you give UnityTransport an invalid address (IP or FQDN) that the transport fails to start?

ShadauxCat replied:
What I mean by the async part is that, when hostname resolution is available, we won't know if a hostname is valid or not until the DNS resolution step is completed. We're not blocking on that step, so yes, StartClient and StartServer are non-async but they won't have the information yet to be able to return false. We won't find out until we get a "failed to connect" result later on in the process.

NoelStephensUnity committed:
test and updates

Adding FQDN validation and updating log errors provided to users.
Adding back the invalid address tests to handle newer and older versions of UTP.

NoelStephensUnity committed:
style

removing comment.

NoelStephensUnity committed:
test - fix

Forgot to update for one of the newer error messages when using an invalid address.

NoelStephensUnity committed:
test - update

Fixing an issue with `ClientDisconnectMultipleClients` beginning to disconnect clients before they were connected.
Added `UnityTransportTestHelpers.WaitForMultipleNetworkEvents` helper method to wait for a specific number of events of a specified type.
Added a `UnityTransportTestHelpers.VerboseDebug` helper method.

NoelStephensUnity committed:
style

renaming the verbose debug flag to VerboseDebug

NoelStephensUnity committed:
style

removing VerboseDebug check on a VerboseLog call.

NoelStephensUnity reviewed (approved):
Added a few updates to testing, testing helpers, and updated error message when using a bad IP and/or FQDN.

NoelStephensUnity committed:
test

increasing disconnect timeout for 1st client  during the ClientDisconnectMultipleClients to account for a slower than expected system.

NoelStephensUnity committed:
style - PVP

Working around PVP issue.

NoelStephensUnity committed:
fix

Re-arranging parameters between both methods.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -256,10 +259,14 @@ private static NetworkEndpoint ParseNetworkEndpoint(string ip, ushort port, bool
                     !NetworkEndpoint.TryParse(ip, port, out endpoint, NetworkFamily.Ipv6))
                 {
+#if HOSTNAME_RESOLUTION_AVAILABLE && UTP_TRANSPORT_2_4_ABOVE
+                    return default;
```
Do we want to respect the silent parameter?

NoelStephensUnity replied:
Actually you make a good point. _(good catch)_
The client uses this method once, but the server uses it in a few places. 
It needs a bit of a refactor... perhaps a bool signifying whether it is parsing it for a server address, which in that case we should honor the silent flag, or if it is a client... which in that case we might do a Fqdn check if it reaches that point.

NoelStephensUnity committed:
update

Refactoring the end point parsing within ConnectionAddressData.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into feat/connect_using_hostname_2.0.2
